OLIVE-19 Enable automatic provisioning of jenkins instance dedicated to running analytics tasks.

jbzdak opened the pull request:
# Testing instructions 

00. Create a virtualenv and install ``requirements.txt`` there. 
0. Verify that you have access to: https://github.com/edx-ops/edx-jenkins-job-dsl/
1. Generate a new ssh-key. 
2. Attach this ssh-key to your github account. 
3. Create a yaml file containing configuration, required configuration is explained in ``Readme.md`` (if it's not obvious what to put in the file it's a documentation bug :))
4. Go to ``vagrant/base/analytic_jenkins`` and issue ``VAGRANT_JENKINS_LOCAL_VARS_FILE=your/path vagrant up``
5. Expected result is that you will get a working jenkins istance that you'll be able to log in to ``192.168.33.15:8080`` or ``localhost:8080``. 
6. You'll be able to login using ``jenkins:password you set`` credentials to the instance. 
7. There will be a role called ``seed`` which role was already ran single time with success result.
8. During the run ``seed`` role created a new role called: ``stub-analytics-jobs``. 
9. Remove the key you added to github in step 2. :) 

# Documentation check 

1. Check documentation of the new role ``playbooks/roles/jenkins_analytics/README.md`` to check whether you understand how to configure the this role. 
2. Check outstanding issues to check which ones are grievous enough to fix right now :)

edx-webhook commented:
Thanks for the pull request, @jbzdak! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=2810&repo=edx%2Fconfiguration

jbzdak commented on playbooks/roles/jenkins_master/defaults/main.yml:
```diff
@@ -34,7 +34,7 @@ jenkins_plugins:
     - { name: "github", version: "1.14.0" }
     - { name: "github-api", version: "1.69" }
     - { name: "github-oauth", version: "0.20" }
-    - { name: "github-sqs-plugin", version: "1.6" }
```
Version 1.6 wasn't present in the repositories.

e0d replied:
@jzoldak any context on the missing version?

jzoldak replied:
tl;dr :+1: to this change

We're still using my fork (see the jenkins_custom_plugins section below in the same file) on build.jenkins.edx.org because the new feature that we added was accepted upstream but the author still hasn't made an official release.

jbzdak commented on playbooks/roles/jenkins_master/tasks/main.yml:
```diff
@@ -99,7 +99,7 @@
     owner: "{{ jenkins_user }}"
     group: "{{ jenkins_group }}"
-    mode: 644
+    mode: "644"
```
It looked like something (yaml, ansible?) treated 644 as octal number while reading and converting it to decimal on output (resulting in messed permissions). I just changed mode to string, which solved the problem.

e0d commented on playbooks/roles/jenkins_analytics/defaults/main.yml:
```diff
@@ -0,0 +1,8 @@
+  name: unix
+  service: su
+  plain_password: jenkins
+  crypted_password: $6$rAVyI.p2wXVDKk5w$y0G1MQehmHtvaPgdtbrnvAsBqYQ99g939vxrdLXtPQCh/e7GJVwbnqIKZpve8EcMLTtq.7sZwTBYV9Tdjgf1k.
```
There should be some documentation here regarding the appropriate use of this default password.  Typically, we try to indicate in default passwords that they are to be overridden.  Also, is it possible to crypt the password in the role?  If not include docs describing the process.

jbzdak replied:
@e0d  I'll document that whole ``jenkins_auth_realm`` should be overriden. Crypting remotely is possible but I'll investigate if crypting remotely results in 'cleaner' solution

jbzdak commented:
@itsjeyd This should be ready for a initial review. When you get access to https://github.com/edx-ops/edx-jenkins-job-dsl/ you'll be able to just vagrant up, for now just look for any things that should be polished, and check: ``Known issues`` section in ``Readme.md`` and check if I'm not too cautious.

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+A role that sets-up Jenkins for deploying analytics stack
+on AWS EMR.
+
+It does following steps:
```
@jbzdak "It performs the following steps" maybe? If you change this, you can change the verbs in the items below to "Install", "Configures", "Enable".

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+
+It does following steps:
+
+* Installs jenkins using ``jenkins_master``
```
@jbzdak "jenkins" ⟶ "Jenkins" (check remaining text for other occurrences of this)

Also, single backticks should suffice to get GitHub's Markdown renderer to display `jenkins_master` as code. When you change this, make sure to make the use of backticks consistent in the entire file.

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+You need to provide a json file containing credentials
+to be uploaded to server. This file is **a local file**
+suitable for use in ``copy`` module. This file is set by
+following variable:
```
@jbzdak "The following variable specifies the location of this file:"

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+    jenkins_credentials_file:  files/ansible-jenkins-credentials.json
+
+
+Used by command waiting on jenkins start-up after running ``jenkins_master``
```
@jbzdak "Used by" ⟶ "Variables used by"

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+Unix realm requires following settings:
+
+``service`` --- jenkins uses PAM configuration for this service. `su` is
+a safe choice as it doesn't require that user has ability to login remotely/
```
@jbzdak Replace trailing `/` with a period.

"require that user has ability" ⟶ "require a user to have the ability"

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+### Seed job configuration
+
+Seed job is configured in ``jenkins_seed_job`` variable, which has following
+attributes:
```
@jbzdak As general comment, could you turn the list of attributes below into an actual list? It should look like this:

```
* `name`: Name of the job in Jenkins

* ...

* `scm`: ... It has the following properties:

  * `scm.type`: ...
```

Rendered version:

* `name`: Name of the job in Jenkins

* ...

* `scm`: ... It has the following properties:

  * `scm.type`: ...

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+
+``name`` Name of the job in Jenkins.
+
+``time_trigger``: A jenkins cron entry defining how often this job is ran.
```
@jbzdak You're using `:` here to separate the attribute name from the description. This looks great, can you do that for the remaining items as well?

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+``removed_view_action`` What to do when view created by the seed job is removed.
+This can be either  ``DELETE`` or``IGNORE``.
+``scm`` Scm object is used to define from where to download
+the jobs. It has following properties:
```
@jbzdak For consistency, please add an empty line between the items above.

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+``scm`` Scm object is used to define from where to download
+the jobs. It has following properties:
+
+``scm.type`` For today it must have value of ``git``.
```
@jbzdak "For today" ⟶ "As of today"

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+
+``scm.type`` For today it must have value of ``git``.
+
+``scm.url`` Url for the repository
```
@jbzdak "Url" ⟶ "URL"

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+
+``scm.credential_id`` Id of credential used to authenticate to the repository.
+
+``scm.targed_jobs`` A shell glob expression relative to repo selecting jobs
```
@jbzdak "targed" is probably a typo? (There is another instance of this in the "Example scm configuration" below.)

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+to import
+
+``scm.additional_classpath`` A path relative to repo root, pointing to a path
+that contains additional groovy scripts used by the seed jons.
```
@jbzdak "jons" ⟶ "jobs"

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+        cat file.yaml |  python -c "import yaml,json,sys; print(json.dumps(yaml.load(sys.stdin)))" > file.json
+
+
+See ``jenkins_analytics/meta/ansible-jenkins-credentials.yaml`` a example
```
@jbzdak "a example configuration file format" ⟶ "for an example configuration file"

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+------------
+
+1. SSH keys without password don't work.
+2. This playbook should have been a module: ``_execute_ansible_cli.yaml``
```
@jbzdak What does that mean? (It might not be necessary to go into more detail about this here, for now I'm just asking because I don't know what to make of this item.)

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+2. This playbook should have been a module: ``_execute_ansible_cli.yaml``
+3. Anonymous user has discover and get job permission, as without it
+  ``get-job``, ``build <<job>>`` commands wouldn't work.
+  Giving anonymous these permissions a workaround for
```
@jbzdak "permissions a workaround" ⟶ "permissions is a workaround"

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+  ``get-job``, ``build <<job>>`` commands wouldn't work.
+  Giving anonymous these permissions a workaround for
+  transient Jenkins issue (reported [couple][1] [of][2] [times][3]).
+4. We force unix authentication method --- that is every user that can login
```
@jbzdak "---" ⟶ "--"

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,139 @@
+Dependencies
+------------
+
+- jenkins_master
```
@jbzdak Missing backticks around `jenkins_master`.

itsjeyd commented on vagrant/base/analytic_jenkins/Vagrantfile:
```diff
@@ -0,0 +1,37 @@
+# -*- mode: ruby -*-
+# vi: set ft=ruby :
```
@jbzdak There seems to be one other Vagrantfile that includes this line, but I think it would be better to leave out editor-specific customizations in these files :)

itsjeyd commented on vagrant/base/analytic_jenkins/Vagrantfile:
```diff
@@ -0,0 +1,37 @@
+  # https://docs.vagrantup.com.
+
+  # Every Vagrant development environment requires a box. You can search for
+  # boxes at https://atlas.hashicorp.com/search.
```
@jbzdak Leftover comment, please remove.

itsjeyd commented on vagrant/base/analytic_jenkins/Vagrantfile:
```diff
@@ -0,0 +1,37 @@
+
+   config.vm.provision :ansible do |ansible|
+    ansible.playbook = "../../../playbooks/analytics-jenkins.yml"
+    # ansible.verbose = "vvvv"
```
@jbzdak For a base Vagrantfile, `vvvv` is useful I think; no need to comment it out in the final version :)

itsjeyd commented on playbooks/roles/jenkins_analytics/defaults/main.yml:
```diff
@@ -0,0 +1,32 @@
+---
+
+# See Readme.md for variable descriptions
```
@jbzdak "Readme.md" ⟶ "README.md"

itsjeyd commented on playbooks/roles/jenkins_analytics/defaults/main.yml:
```diff
@@ -0,0 +1,32 @@
+
+# See Readme.md for variable descriptions
+
+jenkins_credentials_file:  files/ansible-jenkins-credentials.json
```
@jbzdak I don't think you need the `files/` prefix here. Ansible should look for the file in that folder automatically.

jbzdak replied:
Technically it's just an example, for reasons that are not really clear to me ansible wanted files prefix when I launched it from vagrant. But that's not really important, and I'll just move local credentials and fix this so everything is clearer.

itsjeyd commented on playbooks/roles/jenkins_analytics/defaults/main.yml:
```diff
@@ -0,0 +1,32 @@
+  crypted_password: $6$rAVyI.p2wXVDKk5w$y0G1MQehmHtvaPgdtbrnvAsBqYQ99g939vxrdLXtPQCh/e7GJVwbnqIKZpve8EcMLTtq.7sZwTBYV9Tdjgf1k.
+
+# To include provisioning many seed jobs you'd need to use
+# Ansible 2.0+, and then include + with_items that was re-introduced then
```
@jbzdak I'm having a bit of a hard time with this comment. What do these mean:

- "To include provisioning many seed jobs"
- "and then include + with_items that was re-introduced"

jbzdak replied:
I'll rephrase it. Basically what I'm trying to say is: that having many seed jobs would be very easy to do in ansible <=1.6 and >=2.0, and rather hard for Ansible in between.

itsjeyd commented on playbooks/roles/jenkins_analytics/defaults/main.yml:
```diff
@@ -0,0 +1,32 @@
+# To include provisioning many seed jobs you'd need to use
+# Ansible 2.0+, and then include + with_items that was re-introduced then
+jenkins_seed_job:
+  name: seed
```
@jbzdak Could you make the name a bit more descriptive?

itsjeyd commented on playbooks/roles/jenkins_analytics/defaults/main.yml:
```diff
@@ -0,0 +1,32 @@
+
+
+
+
```
@jbzdak One final newline per file should be enough :)

itsjeyd commented on playbooks/roles/jenkins_analytics/meta/ansible-jenkins-credentials.yaml:
```diff
@@ -0,0 +1,30 @@
+# Note for some reason ssh keys without passwords dont work.
+
+credentials:
+# id is a scope-unique credentiail identifier
```
@jbzdak "credentiail" ⟶ "credential" (another occurrence below)

Can you please indent the comments (YAY comments! :)) to the same level as the things they are describing?

itsjeyd commented on playbooks/roles/jenkins_analytics/meta/ansible-jenkins-credentials.yaml:
```diff
@@ -0,0 +1,30 @@
+    username: jbzdak
+# Type of credential, see other entries for example
+    type: ssh-private-key
+# Pasphrase, ssh keys don't work without passphrase
```
@jbzdak "Pasphrase" ⟶ "Passphrase"

itsjeyd commented on playbooks/roles/jenkins_analytics/meta/ansible-jenkins-credentials.yaml:
```diff
@@ -0,0 +1,30 @@
+      DEK-Info: AES-128-CBC,....
+
+      Key conents
+      -----END RSA PRIVATE KEY-----
```
@jbzdak I'm not sure if `meta` is the right place for this file. `files` would probably be more appropriate (and more convenient, since that's where the role expects to find the JSON version of it).

jbzdak replied:
@itsjeyd Its an example file, not a template, nor config file. The real file can't be checked to the public repository, as it contains a private key and password :) I could paste this to readme but then readme would have too much code snippets, to my taste.

jbzdak replied:
It's just an example file, probably I could move it to readme, but there is already too much code in readme for my taste, so I moved this there, could create a ``doc`` sub-folder, but it might not make sense.

itsjeyd replied:
@jbzdak I agree that this file is kind of tricky to place, and putting it in `files` also feels a little weird. But since `meta` is for dependencies, I think it would be good if we could move it somewhere else. I like your suggestions about adding the contents to the README and getting rid of the file, I think it might be the best option.

itsjeyd commented on playbooks/roles/jenkins_analytics/meta/main.yml:
```diff
@@ -0,0 +1,2 @@
+dependencies:
+ - jenkins_master
```
@jbzdak Use two spaces for indentation here.

itsjeyd commented on playbooks/roles/jenkins_analytics/tasks/_add_credentials.yaml:
```diff
@@ -0,0 +1,22 @@
+- name: upload credentials file
+  copy: src={{ jenkins_credentials_file }} dest={{ jenkins_credentials_file_dest }}
+        mode="600"
+        owner={{jenkins_user}}
```
@jbzdak Missing spaces around variable names.

itsjeyd commented on playbooks/roles/jenkins_analytics/tasks/_add_credentials.yaml:
```diff
@@ -0,0 +1,22 @@
+    jenkins_command_string: "groovy {{ jenkins_credentials_script }}"
+- name: clean up
+  file: name={{ jenkins_credentials_root }} state=absent
+
```
@jbzdak To improve readability, could you insert an empty line after each task? This also applies to the other files in the `tasks` folder.

There are two trailing newlines here, please remove one of them.

itsjeyd commented on playbooks/roles/jenkins_analytics/tasks/_auth_realm_unix.yaml:
```diff
@@ -0,0 +1,26 @@
+  changed_when: false
+- command: adduser {{ jenkins_user }}  shadow
+  when: out.rc != 0
+  notify: restart Jenkins
```
@jbzdak If you used the `user` module for these customizations, would you really be recreating the `jenkins_user` from scratch? I'm wondering if you couldn't use the `groups` option of that module to add `shadow`. (A similar comment applies to the code where you're setting the password below.)

jbzdak replied:
I thought that user module was supposed to set all user properties, and but after reading the sources it seems that I can update user's password and add him to a group without touching things set by ``jenkins_master`` so I'll do it.

itsjeyd commented on playbooks/roles/jenkins_analytics/tasks/_auth_realm_unix.yaml:
```diff
@@ -0,0 +1,26 @@
+  notify: restart Jenkins
+- name: template config.xml
+  template: src=jenkins.config.main.xml dest={{ jenkins_home }}/config.xml
+            owner={{ jenkins_user }} group={{ jenkins_group }}
```
@jbzdak Could you please format the module arguments according to [this specification](https://openedx.atlassian.net/wiki/display/OpenOPS/Ansible+Code+Conventions#AnsibleCodeConventions-Formatting)? There are more instances of this in the remaining files.

itsjeyd commented on playbooks/roles/jenkins_analytics/tasks/_create_seed_job.yaml:
```diff
@@ -0,0 +1,34 @@
+- set_fact:
+    temporary
```
@jbzdak What does this do? (Just asking, not saying you should remove it)

jbzdak replied:
Nothing, should be removed :)

itsjeyd commented on playbooks/roles/jenkins_analytics/tasks/_create_seed_job.yaml:
```diff
@@ -0,0 +1,34 @@
+    get_job_output: "{{ jenkins_command_output }}"
+- debug: var=jenkins_command_output
+- name: create a seed job
+
```
@jbzdak This blank line shouldn't be here I think.

itsjeyd commented on playbooks/roles/jenkins_analytics/tasks/_create_seed_job.yaml:
```diff
@@ -0,0 +1,34 @@
+    jenkins_command_prefix: "cat /tmp/{{ job_name }} | "
+  when: get_job_output.rc != 0
+
+- name: create a seed job
```
@jbzdak This has the same name as the task above, please update.

itsjeyd commented on playbooks/roles/jenkins_analytics/tasks/_execute_ansible_cli.yaml:
```diff
@@ -0,0 +1,44 @@
+---
+# tasks file for jenkins_cli
+- fail: msg=for now we can execute commands iff jenkins auth relm is unix
```
@jbzdak "relm" ⟶ "realm"

itsjeyd commented on playbooks/roles/jenkins_analytics/tasks/_execute_ansible_cli.yaml:
```diff
@@ -0,0 +1,44 @@
+    {{ jenkins_command_prefix|default('') }} java -jar {{ jenkins_cli_jar }} -s http://localhost:{{ jenkins_port }}
+    {{ jenkins_command_string }}
+  register: jenkins_command_output
+  ignore_errors: "{{ jenkins_ignore_cli_errors| default (False) }}"
```
@jbzdak `jenkins_ignore_cli_errors| default (False)` ⟶ `jenkins_ignore_cli_errors|default(False)`

itsjeyd commented on playbooks/roles/jenkins_analytics/tasks/_execute_ansible_cli.yaml:
```diff
@@ -0,0 +1,44 @@
+  file: name=jenkins_cli_root state=absent
+
+- name: clean up
+  command: rm -rf $HOME/.jenkins
```
@jbzdak Maybe calling this file "_execute_jenkins_cli.yaml" would be more fitting?

jbzdak replied:
Of course!

itsjeyd commented on playbooks/roles/jenkins_analytics/tasks/main.yml:
```diff
@@ -0,0 +1,19 @@
+---
+# Jenkins is started by the trigger of jenkins_master job, which trigger
+# is executed after whole playbook. Since we expect Jenkins running
+# force it to be started now
```
@jbzdak `jenkins_analytics` has `jenkins_master` as a dependency -- doesn't this mean that any handlers that are notified by `jenkins_master` would run before any tasks that belong to `jenkins_analytics`? Not suggesting to change anything here just yet -- asking because I'm not sure.

jbzdak replied:
It actually failed because Jenkins wasn't started, so no. Docs say that: 
```
These ‘notify’ actions are triggered at the end of each block of tasks in a playbook, and will only be triggered once even if notified by multiple different tasks.
```
So nothing about roles. However I learned that I can add following: 

```
- meta: flush_handlers
```

to force all handlers to run

jbzdak replied:
After some thought I'd leave the service solutions. Handlers are not launched between roles, and without ``- service: name=jenkins state=started``, due to some transient error you can end up in a state where jenkins is stopped becaouse handler that was supposed to start it didn't fire, and then this would allways fail.

e0d replied:
See my comment about handlers above.

jbzdak replied:
@e0d These handlers are part of role I depend on, and I'd rather not modify not essential things at this stage. I think that we might be able to do some additional cleanups to follow-up pr. 

Problem what that due to some transient error that occurred at the wrong moment, the build was broken until restart or someone manual starts the Jenkins.

itsjeyd commented on playbooks/roles/jenkins_analytics/tasks/main.yml:
```diff
@@ -0,0 +1,19 @@
+#- include: _execute_ansible_cli.yaml
+#  vars:
+#    jenkins_command_string: "who-am-i"
+#- debug: var=jenkins_command_output
```
@jbzdak Leftover TODO

itsjeyd commented on playbooks/roles/jenkins_analytics/templates/addCredentials.groovy:
```diff
@@ -0,0 +1,76 @@
+
+d.credentials.each { cred ->
+
+    //def scope = "GLOBAL"
```
@jbzdak Leftover comment

itsjeyd commented on playbooks/roles/jenkins_analytics/templates/jenkins-pass-file:
```diff
@@ -0,0 +1 @@
+{{jenkins_auth_realm.plain_password}}
```
@jbzdak Missing spaces around variable name. Please check for this in the remaining files you added as well, and update them if necessary.

itsjeyd commented on vagrant/base/analytic_jenkins/Vagrantfile:
```diff
@@ -0,0 +1,30 @@
+    ansible.extra_vars = {}
+  end
+
+end
```
@jbzdak I missed this earlier: The parent directory of this file should be called "analytics_jenkins" (it's currently missing an "s"). You'll also need to update the test instructions in the description of the PR.

itsjeyd commented on playbooks/roles/jenkins_analytics/defaults/main.yml:
```diff
@@ -0,0 +1,27 @@
+  crypted_password: $6$rAVyI.p2wXVDKk5w$y0G1MQehmHtvaPgdtbrnvAsBqYQ99g939vxrdLXtPQCh/e7GJVwbnqIKZpve8EcMLTtq.7sZwTBYV9Tdjgf1k.
+
+# For now only a single seed job is supported, adding more would require
+# Ansibe 2.+ or converting _execute_jenkins_cli to a module
```
@jbzdak "Ansibe" ⟶ "Ansible"

This is much clearer now, thanks!

itsjeyd commented on playbooks/roles/jenkins_analytics/tasks/_add_credentials.yaml:
```diff
@@ -0,0 +1,27 @@
+- name: create credentials dir
+  file: name={{ jenkins_credentials_root }} state=directory
+
+- name: upload groovy scipt
```
@jbzdak "scipt" ⟶ "script" (missed that earlier, sorry)

itsjeyd commented on playbooks/roles/jenkins_analytics/tasks/_add_credentials.yaml:
```diff
@@ -0,0 +1,27 @@
+- name: upload credentials file
+  copy: src={{ jenkins_credentials_file }} dest={{ jenkins_credentials_file_dest }}
+        mode="600"
+        owner={{ jenkins_user }}
```
@jbzdak Could you adjust the formatting of the module arguments for this task and the "upload groove script" task above as well?

Note that when you use multiple lines, the coding standards suggest to use this format:

```
- file:
    dest: "{{ test }}"
    src: "./foo.txt"
```

jbzdak replied:
OK, I didn't notice that they wanted dictionaries there.

itsjeyd commented on playbooks/roles/jenkins_analytics/tasks/main.yml:
```diff
@@ -0,0 +1,14 @@
+  vars:
+    job_name: "{{ jenkins_seed_job.name }}"
+    job: "{{ jenkins_seed_job }}"
+
```
@jbzdak The formatting looks so much better now! Could you add vertical space between tasks and includes in this file as well? (And remove one trailing newline at the end of the file.)

jbzdak replied:
Done

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,162 @@
+This  file is **a local file** suitable for use in `copy` module.
+The following variable specifies the location of this file:
+
+    jenkins_credentials_file:  files/ansible-jenkins-credentials.json
```
@jbzdak Please remove `files/` prefix here as well.

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,162 @@
+  * `scm.target_jobs` A shell glob expression relative to repo selecting
+    jobs to import.
+  * `scm.additional_classpath` A path relative to repo root, pointing to a
+     directory that contains additional groovy scripts used by the seed jobs.
```
@jbzdak To be consistent with the previous items, please add `:` to the `scm` entries as well (e.g. "`scm.type`: As of today ...").

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,162 @@
+
+        cat file.yaml |  python -c "import yaml,json,sys; print(json.dumps(yaml.load(sys.stdin)))" > file.json
+
+Example credential file is here (in yaml` format):
```
@jbzdak Remove leftover backtick.

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,162 @@
+        username: jbzdak
+        type: username-password
+        description: Autogenerated by ansible
+        password: 'foobarbaz'
```
@jbzdak I think you can remove the first "section" (starting in L114) from this file. When I tested the PR I based my YAML file on the example provided in the PR description, and that only contains the second "section" starting below.

jbzdak replied:
It works without the password, as no password is seeded to launch a seed job. However this PR is only a start of work, and since adding support of passwords didn't require any work from me (initial version of ``addCredentials.groovy`` already contained it I decided to document this feature). So this example is just a documentation of potentially useful feature.

itsjeyd commented:
@jbzdak :+1: once the build is green (failure seems unrelated to your changes).

jbzdak commented:
Please review @e0d @mulby @brianhw . This PR implements OLIVE-19 story. The TravisCi is red but I believe that this is unrelated, so the best course of action is to wait for someone to fix the underlying issue and then the cherry pick the fix. If that's not the case I'll be happy to fix it of course.

e0d commented on playbooks/roles/jenkins_analytics/handlers/main.yml:
```diff
@@ -0,0 +1,3 @@
+---
+- name: restart Jenkins
```
We've generally been avoiding handlers.  I find that they add indirection without much value compared to say explicitly restarting services in tasks.  In this case, as we are not deploying AMIs on upgrade, thus keeping servers around longer, I'm open to it if you think it adds value.

jbzdak replied:
I think that in this case it makes sense, this handler is used when updating Jenkins users. I could just restart it always, and this would introduce a slight delay for each re-provisioning of the server (time needed for Jenkins to start itself before seed job can be updated). In the long run I think that this handler will be used more and more, as new features are added to this role, and there will be more tasks where we need to restart Jenkins only after settings got updated.

jbzdak replied:
I have mixed feelings about handlers here. In current solution I have to explicitly launch all pending handlers before running CLI, but without them I'd either have restart Jenkins unconditionally at that stage or  have to do something like that: 

```
- user:
    name: "{{ jenkins_user }}"
    groups: shadow
    append: yes
    password: "{{ jenkins_auth_realm.hashed_password }}"
    update_password: always
  notify: restart Jenkins
  register: user_out

- name: template config.xml
  template:
    src: jenkins.config.main.xml
    dest: "{{ jenkins_home }}/config.xml"
    owner: "{{ jenkins_user }}"
    group: "{{ jenkins_group }}"
  notify: restart Jenkins
  register: config_out

- name: restart jenkins
  service: name=jenkins state=restarted 
  when: user_out.changed or config_out.changed 

```

e0d replied:
What you've written looks reasonable, but I don't think restarting always would be the end of the world.  For example, would we upgrade Jenkins in place now that we can build it from scratch consistency with one command?  If we end up essentially managing the service in a red/black fashion, there's nothing to optimize here.

jbzdak commented on playbooks/roles/jenkins_analytics/tasks/_create_seed_job.yaml:
```diff
@@ -0,0 +1,33 @@
+  include: _execute_jenkins_cli.yaml
+  vars:
+    jenkins_command_string: "build {{ job_name }} -s"
+    jenkins_command_prefix: "sleep 30 && "
```
This sleep is here most probably because there was a delay between completion of ``update-job`` and jenkins being ready to run this job. Need to recheck and remove this.

bradenmacdonald replied:
ping - reminder to investigate and remove this `sleep` or replace it with something better.

bradenmacdonald commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,163 @@
+
+### Seed job configuration
+
+Seed job is configured in `jenkins_seed_job` variable, which has following
```
Is this variable meant to be customizable? If so, it should be uppercase, shouldn't it? https://openedx.atlassian.net/wiki/display/OpenOPS/Ansible+Code+Conventions

jbzdak commented on vagrant/base/analytics_jenkins/Vagrantfile:
```diff
@@ -0,0 +1,35 @@
+PRIVATE_IP = ENV['VAGRANT_PRIVATE_IP'] || '192.168.33.15'
+
+Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
+  config.vm.box = 'ubuntu/trusty64'
```
Ruby linter complained that `""` string should be reserved only for interpolated strings.

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,196 @@
+# Jenkins Analytics
+
+A role that sets-up Jenkins for deploying analytics stack on AWS EMR.
```
@jbzdak "sets-up" ⟶ "sets up"

bradenmacdonald replied:
Is Jenkins "deploying **the** analytics stack" or "scheduling analytics tasks"? (Or both)

jbzdak replied:
@bradenmacdonald  I'll go with "scheduling analytics tasks", @pomegranited may amend this description to match awesome things her DSL jobs do :)

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,196 @@
+all required variables from this section.
+
+This file needs to contain, at least, following variables these 
+variables are explained in the next few sections. 
```
@jbzdak "This file needs to contain, at least, the following variables (see the next few sections for more information about them):"

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,196 @@
+#### Jenkins credentials
+
+Jenkins contains it's own credential store, to fill it with credentials 
+please use ``JENKINS_ANALYTICS_CREDENTIALS`` variable. This variable 
```
@jbzdak "it's" ⟶ "its"

Nit: Use single backticks around `JENKINS_ANALYTICS_CREDENTIALS`.

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,196 @@
+for now passwords and ssh-keys are supported. 
+
+If you only need credentials to access github repositories
+you can override `JENKINS_ANALYTICS_GITHUB_KEY`, with a 
```
@jbzdak With a *what*? :)

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,196 @@
+credential for given task. 
+
+
+Exampls of credentials variables:
```
@jbzdak "Exampls" ⟶ "Examples"

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,196 @@
+          Proc-Type: 4,ENCRYPTED
+          DEK-Info: AES-128-CBC,....
+
+          Key conents
```
@jbzdak "conents" ⟶ "contents"

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,196 @@
+          Key conents
+          -----END RSA PRIVATE KEY-----
+
+### All configuration
```
@jbzdak If this complements "End-user editable configuration", maybe change it to something like "General configuration"?

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,196 @@
+  This can be either  `DELETE` or`IGNORE`.
+* `removed_view_action`: What to do when view created by the seed job is removed.
+  This can be either  `DELETE` or`IGNORE`.
+* `scm` Scm object is used to define seed job repository and related settings.
```
@jbzdak Missing colon after `scm`

itsjeyd commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,196 @@
+  * `scm.url`: URL for the repository.
+  * `scm.credential_id`: Id of credential used to authenticate to the repository.
+    It is optional, if missing or falsy credentials will be omitted. Please note
+    that when you use ssh repository url, you'll need to set up key irregardles 
```
@jbzdak "This setting is optional. If it is missing or falsy, credentials will be omitted."

"irregardles" ⟶ "irregardless"

itsjeyd commented on playbooks/roles/jenkins_analytics/defaults/main.yml:
```diff
@@ -0,0 +1,40 @@
+  service: su
+  # Change this default password: (see README.md to see how you can do it)
+  plain_password: "{{JENKINS_ANALYTICS_USER_PASSWORD_PLAIN}}"
+  crypted_password: "{{JENKINS_ANALYTICS_USER_PASSWORD_CRYPTED}}"
```
@jbzdak Missing spaces around variable names here.

itsjeyd commented on vagrant/base/analytics_jenkins/Vagrantfile:
```diff
@@ -0,0 +1,35 @@
+  end
+
+  unless ENV['VAGRANT_JENKINS_LOCAL_VARS_FILE']
+    raise 'Please set VAGRANT_JENKINS_LOCAL_VARS_FILE enviorment varible, '\
```
@jbzdak "enviorment" ⟶ "environment"

brianhw replied:
nit:  "varible, that" => "variable.  That"

itsjeyd commented on playbooks/roles/jenkins_analytics/templates/addCredentials.groovy:
```diff
@@ -0,0 +1,90 @@
+}
+
+def jsonSlurper = new JsonSlurper()
+def d = jsonSlurper.parse(new FileReader(jsonFile))
```
@jbzdak I don't know if this variable is named according to some convention, but if not, could we use a more descriptive variable name here?

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,196 @@
+
+A role that sets-up Jenkins for deploying analytics stack on AWS EMR.
+
+This role performs following steps:
```
nit: "performs the following steps:"

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,196 @@
+* Installs Jenkins using `jenkins_master`
+* Configures `config.xml` to enable security and use
+  Linux Auth Domain.
+* Enable the use of jenkins clis
```
nit: use "Enables", to match other bullets.  Also capitalize Jenkins, and all-caps CLI acronym.  And add a period to the end of each of these, as you did in the next one.

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,196 @@
+## Configuration
+
+When you are using vagrant you **need** to set `VAGRANT_JENKINS_LOCAL_VARS_FILE`
+environment variable, this variable must point to a file containing 
```
nit:  "variable.  This variable"...

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,196 @@
+You'll need to override default jenkins user password, please do that
+as this sets up the **shell** password for this user. 
+
+You'll need to set both plain password and crypted one, 
```
nit:  "a plain password and a crypted one.  To obtain...."

bradenmacdonald replied:
Nit: Is "crypted" a word that Jenkins forces us to use? AFAIK it's not a proper word (it actually means "vaulted" as in a ceiling acc. to Webster), and I'd much prefer "encrypted" or "hashed" if we have a choice.

jbzdak replied:
@bradenmacdonald Crypted seems to take it's origin in [man 3 crypt](http://man7.org/linux/man-pages/man3/crypt.3.html), and means password encrypted using this function in a format that this function produces.

bradenmacdonald replied:
@jbzdak The "man 3 crypt" link you sent does not contain the word "crypted" anywhere :-p

"Encrypted" or "hashed" are definitely more appropriate words.

jbzdak replied:
@bradenmacdonald  After some brief research I found word crypted in the [python docs](https://docs.python.org/2/library/crypt.html) ;) But I agree that hashed is more correct in this context.

brianhw commented on playbooks/roles/jenkins_analytics/defaults/main.yml:
```diff
@@ -0,0 +1,40 @@
+JENKINS_ANALYTICS_CREDENTIALS:
+  - id: github-deploy-key
+    scope: GLOBAL
+    username: jbzdak
```
Is this a good default? : )

jbzdak replied:
Very good catch!

bradenmacdonald replied:
Note "username: jbzdak" appears several times in this PR's diff. I'm not clear on why you need the GitHub username at all actually? SSH cloning from github always uses the same username (`git`), and for something like this we should be using GitHub deploy keys, not individual user's private keys.

jbzdak replied:
Jenkins credentials plugin requires username for private keys, this username appears to be ignored by git plugin (which makes sense). So I'll change username to git in every example just to match intended user. Possibly this will be used by plugins that log somewhere via SSH, so I can just remove username property.

brianhw commented on playbooks/roles/jenkins_analytics/tasks/_add_credentials.yaml:
```diff
@@ -0,0 +1,31 @@
+  template:
+    src: addCredentials.groovy
+    dest: "{{ jenkins_credentials_script }}"
+    mode: "600"
```
does this need an owner?

brianhw commented on playbooks/roles/jenkins_analytics/templates/addCredentials.groovy:
```diff
@@ -0,0 +1,90 @@
+ * For a given json file, it will create credential in a Jenkins instance.  The script can be run safely
+ * multiple times and it will update a credential.
+ *
+ * This is useful in conjunction with the job-dsl to bootstrap an barebones Jenkins instance.
```
nit:  "a barebones"

brianhw commented on playbooks/roles/jenkins_analytics/templates/addCredentials.groovy:
```diff
@@ -0,0 +1,90 @@
+ *
+ * java -jar /var/jenkins/war/WEB-INF/jenkins-cli.jar -s http://localhost:8080 groovy addCredentials.groovy
+ *
+ * For a given json file, it will create credential in a Jenkins instance.  The script can be run safely
```
nit: either "create a credential" or "create credentials".

brianhw commented on vagrant/base/analytics_jenkins/Vagrantfile:
```diff
@@ -0,0 +1,35 @@
+    raise 'Please set VAGRANT_JENKINS_LOCAL_VARS_FILE enviorment varible, '\
+          'that variable should point to a file containing variable '\
+          'overrides for analytics_jenkins role. For required overrides '\
+          'see README.md in the analytics_jenkins role folder'
```
nit: end with period.

bradenmacdonald commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,196 @@
+to obtain crypted password use `mkpasswd` command, for example: 
+`mkpasswd --method=sha-512`. (Note: crypted password is required 
+to have clean "changed"/"unchanged" notification for this step 
+in Ansible.) 
```
@jbzdak If you made `JENKINS_ANALYTICS_USER_PASSWORD_SEED` a parameter instead of the encrypted password, you could pass a seed argument to `mkpasswd`. That way you can let ansible call `mkpasswd` and the result would be consistent every time.

Consider this optional, but I mention it since I don't like having to manually specify data twice in different formats, when one is trivially derived from the other.

jbzdak replied:
I actually discussed this with Tim today, this is doable, but I'm nos sure this will make the code cleaner. 

Here is what I came up with (we can move mkpasswd call to remote host but that is all we can d to make it better): 

```
- name: generate Jenkins password
  local_action: shell echo '{{ JENKINS_ANALYTICS_USER_PASSWORD_PLAIN }}' | mkpasswd --password-fd=0 --method=SHA-512 --salt='{{ JENKINS_ANALYTICS_USER_PASSWORD_SALT }}'
  register: mkpasswd_command_out
  sudo: false

- user:
    name: "{{ jenkins_user }}"
    groups: shadow
    append: yes
    password: "{{ mkpasswd_command_out.stdout }}"
    update_password: always
  notify: restart Jenkins
```

jbzdak replied:
@bradenmacdonald Github hid this, but let's continue the discussion.

bradenmacdonald replied:
@jbzdak Do whatever you feel is cleaner for now, and I'll defer to the devops team to say what they'd prefer for this sort of mkpasswd use.

pomegranited commented:
@jbzdak  I was able to merge your changes into my branch, and provision a fresh vagrant install, so I'm happy :+1: 

Looks like the Travis CI is failing for unrelated reasons.

A couple of minor nits, but don't worry about them if there aren't any further changes.

pomegranited commented on playbooks/roles/jenkins_analytics/tasks/_execute_jenkins_cli.yaml:
```diff
@@ -0,0 +1,47 @@
+  template: src=jenkins-pass-file dest={{ jenkins_cli_pass }} mode="600"
+
+- name: Wait for Jenkins to start up before proceeding.
+  shell: "curl -D - --silent http://localhost:{{ jenkins_port }}/cli/"
```
@jbzdak nit: Ansible warns about this : Consider using get_url module rather than running curl

jbzdak replied:
Ansible can be irritating. I used ``uri`` module instead + check for having CLI ready is now cleaner.

pomegranited commented on playbooks/roles/jenkins_analytics/tasks/_auth_realm_unix.yaml:
```diff
@@ -0,0 +1,23 @@
+  when: jenkins_auth_realm.name != "unix"
+
+- fail: msg=Please change default password for jenkins user
+  when: jenkins_auth_realm.plain_password == 'jenkins'
```
@jbzdak This gave me some trouble, when running the playbook on an previously-provisioned Jenkins instance.
I'm not sure why it's happening, but if I change the jenkins unix user password (plain and hashed) in my ``VAGRANT_JENKINS_LOCAL_VARS_FILE`` using the ``mkpasswd`` mechanism, then I get the following error:

``TASK: [jenkins_analytics | login] ********************************************* 
<localhost> REMOTE_MODULE command java -jar /tmp/jenkins-cli/ubuntu/jenkins_cli.jar -s http://localhost:8080 login --username=jenkins --password-file=/tmp/jenkins-cli/ubuntu/jenkins_cli_pass
<localhost> EXEC ['/bin/sh', '-c', 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1457672784.84-35222673854208 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1457672784.84-35222673854208 && echo $HOME/.ansible/tmp/ansible-tmp-1457672784.84-35222673854208']
<localhost> PUT /tmp/tmpBprCm3 TO /home/ubuntu/.ansible/tmp/ansible-tmp-1457672784.84-35222673854208/command
<localhost> EXEC /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=xyzkdoqwjlmbjsdtypgcpnoleawokwqt] password: " -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-xyzkdoqwjlmbjsdtypgcpnoleawokwqt; LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python /home/ubuntu/.ansible/tmp/ansible-tmp-1457672784.84-35222673854208/command; rm -rf /home/ubuntu/.ansible/tmp/ansible-tmp-1457672784.84-35222673854208/ >/dev/null 2>&1'"'"''
failed: [localhost] => {"changed": true, "cmd": ["java", "-jar", "/tmp/jenkins-cli/ubuntu/jenkins_cli.jar", "-s", "http://localhost:8080", "login", "--username=jenkins", "--password-file=/tmp/jenkins-cli/ubuntu/jenkins_cli_pass"], "delta": "0:00:03.678749", "end": "2016-03-11 05:06:28.608998", "rc": 255, "start": "2016-03-11 05:06:24.930249", "warnings": []}
stderr: Bad Credentials. Search the server log for 1944ba23-5c32-4f90-aeac-808fcdeea7c4 for more details.``

You can see it happening on the analytics.sandbox server.  There, I had to disable this fail check to ensure that the playbook would run.

jbzdak replied:
I'll look into it, and fix the underlying problem.

jbzdak replied:
[Fixed here](https://github.com/edx/configuration/commit/c6e10b4d0b0a58a31847341303d129311019503d#diff-b481e83487ef766c6c1e7a4289d524a2R13) please check when you have a moment

pomegranited replied:
@jbzdak Works great now, thank you!  Just one change needed in the [README.md#L24](https://github.com/edx/configuration/pull/2810/files#diff-66aa428c61cffe7679465c81c4bed63dR24)

jbzdak commented:
@e0d, @brianhw: All outstainting issues have been addressed, or commented on. Please take a second look.

e0d commented on playbooks/roles/jenkins_analytics/tasks/main.yml:
```diff
@@ -0,0 +1,22 @@
+- name: start jenkins now
+  service: name=jenkins state=started
+
+- include: _auth_realm.yaml
```
I assume the _ is to indicate that this should only be used as an include.  It's not part of our current idiom.  Can you clarify?

jbzdak replied:
Yes it is supposed to mean, that this file is private to current role (like `_method` in Python). I don't know what is the proper idiom for that in EdX.

e0d commented on playbooks/roles/jenkins_analytics/templates/addCredentials.groovy:
```diff
@@ -0,0 +1,89 @@
+ *
+ * This is useful in conjunction with the job-dsl to bootstrap a barebone Jenkins instance.
+ *
+ * This script will currently fail if the plugins it requires have not been installed:
```
I don't think this is accurate any longer.

jbzdak replied:
I have removed the YAML dependency, but I believe that other plugins are still required --- as we are touching plugin types not Jenkins core types.

e0d commented on playbooks/roles/jenkins_analytics/templates/jenkins.config.main.xml:
```diff
@@ -0,0 +1,74 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<hudson>
+  <disabledAdministrativeMonitors/>
+  <version>1.638</version>
```
Is this jenkins version or XML version?  Feels like this and numExecutors should be variables or dynamicl

jbzdak replied:
This is version of the Jenkins that generated the XML, and is used by Jenkins to handle config files generated from older Jenkinses (at least that's what the docs say). I'd leave this value as a non-templated one, as if ``jenkins_master`` updates to something never this file will be automatically updated to new version when Jenkins starts itself.

pomegranited commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+This file needs to contain, at least, the following variables 
+(see the next few sections for more information about them): 
+
+* `JENKINS_ANALYTICS_USER_PASSWORD_CRYPTED` 
```
@jbzdak Please change this to ``JENKINS_ANALYTICS_USER_PASSWORD_HASHED``

jbzdak replied:
Done.

jbzdak commented:
@e0d @brianhw All comments have been addressed, when you find time please take another look. Last outstanding issue is whether we should use handlers in the `analytics_jenkins` role, and I have a mixed feelings on this (for me handlers  are very slightly better, but I'd be equally happy to remove them). 

If there are any bigger issues I'd be happy to fix them as a follow up PR as this shouldn't be changed too much just to focus work on #2830.

pomegranited commented on playbooks/roles/jenkins_analytics/templates/credentials_file.json:
```diff
@@ -0,0 +1 @@
+{{ JENKINS_ANALYTICS_CREDENTIALS|to_json }}
```
@jbzdak Last last thing.. I think the Travis checks are failing because they run jsonlint over every .json file, and this one doesn't pass, because it's really a jinga template:
https://travis-ci.org/edx/configuration/jobs/115797255#L928

If you rename the file to ``credentials_file.json.j2``, it'll get [jinga checked, instead of json-linted](https://github.com/edx/configuration/blob/jbzdak/olive-19-auto-provision-analytics-devstack/test.mk#L23)

jbzdak replied:
@pomegranited Ach this is another convention I need to learn :). I'll fix this straight away, I was sure that test failure was unrelated as tests locally didn't print any error message and exited with 0 exit code. New version will be uploaded as soon as I verify Jenkins provisions.

e0d commented on playbooks/roles/jenkins_analytics/tasks/_add_credentials.yaml:
```diff
@@ -0,0 +1,31 @@
+---
+- set_fact:
```
Why are you using set_fact here?

jbzdak replied:
To set what amounts to a temporary variable, as far as temporary variables are possible in Ansible. These two can safely be moved to `defaults/main.yaml`.

e0d commented:
I see the point, code reuse, of having  {code}_execute_jenkins_cli.yaml{code} as an include.  However, I don't want to introduce a new idiom "_private_method" in this PR.  We have other examples of included tasks without the leading _

The other cases of this seem to me to add unnecessary indirection and I'd like to see this code in-lined.

{code}
_auth_realm.yaml
_add_credentials.yaml
_create_seed_job.yaml
_auth_realm_unix.yaml
{code}

Ultimately, I'll probably write an Ansible module for running Jenkins CLI commands, but that doesn't need to be part of this PR.

After addressing the comments above I'm :+1:

jbzdak commented:
Thanks for the conditional approval @e0d ! 

I'll drop the `_` prefixes tommorow. 

Here is the logic for separating tasks in this role, please let me try co convince you that this separation adds to the overall clarity, and let me know if I convinced you, and if not I'll merge the files ASAP :) 

**auth_realms**

I wasn't sure whether Unix auth realm is best for this task, and I was wondering if there will be another one implemented (like GitHub realm). If other realms would be implemented you'd just add new file `_auth_realm_foo.yaml` and then add conditional include in `_auth_realm.yaml`. Which would be clean, as far as Ansible permits. 

If you think that there will be no new auth realms added, then I'll happily merge this into main file. 

**_add_credentials and _create_seed_job**

I'd argue that having them separate files adds to overall maintainability and readability: 

1. Each of them has it's own non-trivial logic, that is now encapsulated in it's own file. Most of this logic (for example placement of the `/tmp/credentials`) is invisible to the outside world. 
2. It's very easy to see what `main/tasks.yaml` does right now, as the file consists of three sub tasks, that have limited interactions. When everything gets merged there we'll lose this clarity. 
3. When we merge everything to `tasks/main.yml` file will be about 120-150 lines long, which is longer than about 80% of `tasks/*.yml` in the whole repository, hardly outliner but still would count as long.

brianhw commented on playbooks/roles/jenkins_analytics/tasks/_execute_jenkins_cli.yaml:
```diff
@@ -0,0 +1,51 @@
+    return_content: yes
+    status_code: 200,403
+  register: result
+  until: (result.status is defined) and (result.status == 403) or (results.status == 200)
```
is there any issue with the order of execution for the 'and' and 'or'?

jbzdak replied:
Yes there is. I'll check that.

brianhw commented on playbooks/roles/jenkins_analytics/tasks/main.yml:
```diff
@@ -0,0 +1,22 @@
+---
+# Jenkins is started by the trigger of jenkins_master job, which trigger
+# is executed after whole playbook. Since we expect Jenkins running
```
I don't understand "which trigger is executed after whole playbook".  How about "... job.  This trigger is executed after all playbooks have been executed."  

I also don't understand the second sentence.  If Jenkins is already running, why force it to start?  Do you mean: "Since we expect Jenkins to be running already, force it to be restarted now." ?

jbzdak replied:
The point is (I'll clear this up) that Jenkins is started by a handler in ``jenkins_master`` role, this trigger causes two problems: 

1. If build fails due to transient error Jenkins might not get started, and needs to be started manually, as hanlder is not notified on subsequent runs. 
2. In normal operation Jenkins would be started after this command, at the end of whole playbook. 

So I just force Jenkins to be started.

brianhw commented on playbooks/roles/jenkins_analytics/tasks/main.yml:
```diff
@@ -0,0 +1,22 @@
+- include: _add_credentials.yaml
+
+# Changing the Jenkins credentials could require jenkins being restarted before
+# CLI works, so if any change was needed flush it now
```
nits:   end with period.  s/jenkins/Jenkins/.    
Is this correct:   "Changing the Jenkins credentials could require that Jenkins be restarted before CLI will work.  So if any change was needed, flush it now."

brianhw commented on playbooks/roles/jenkins_analytics/templates/addCredentials.groovy:
```diff
@@ -0,0 +1,89 @@
+    provider = SystemCredentialsProvider.getInstance()
+    provider.getCredentials().add(new UsernamePasswordCredentialsImpl(scope, id, description, username, password))
+    provider.save()
+    println scope
```
naive question:  is this here for debugging or logging?  Is it still needed?  (Same on line 42.)

jbzdak replied:
Yes it is left as a debug tool. I'll remove this.

brianhw commented on playbooks/roles/jenkins_analytics/templates/seed_job_template.xml:
```diff
@@ -0,0 +1,71 @@
+<project>
+  <actions/>
+  <description>
+    Seed job autogenerated by ansible, it will be overriden.
```
nit:  overridden.

brianhw commented on playbooks/roles/jenkins_analytics/tasks/_execute_jenkins_cli.yaml:
```diff
@@ -0,0 +1,51 @@
+- name: clean up
+  file: name=jenkins_cli_root state=absent
+
+- name: clean up
```
shouldn't this have a different name?

jbzdak replied:
Job names don't need to be unique. I'll add better name to serve as a comment.

brianhw commented on playbooks/roles/jenkins_analytics/tasks/_create_seed_job.yaml:
```diff
@@ -0,0 +1,31 @@
+
+- name: upload_job_file
```
any reason for the underscores in this name?  Is this name special?

jbzdak replied:
Nothing special. Will be removed.

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+Known issues
+------------
+
+1. Playbook named `_execute_ansible_cli.yaml`, should be converted to a
```
nit:   s/a/an/.  And next line should end with a period, not a comma.

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+#### Jenkins user password
+
+You'll need to override default jenkins user password, please do that
+as this sets up the **shell** password for this user. 
```
Defaults are provided for this, so does this need to be overridden?  And if so, should there not be default values provided that would work without being overridden?

jbzdak replied:
In the end (#2830) we will not be able to provide defaults for everything, so I decided that password needs to be overriden as a security precaution against lazy people. Remember that Jenkins is really a remote web shell to a server. 

Other variables might be removed if ``edx-ops/edx-jenkins-job-dsl`` is open sourced :)

brianhw replied:
But I don't see where you get people to override the default Jenkins user password.  In playbooks/roles/jenkins_analytics/defaults/main.yml, you provide a default.  Won't that just work?

jbzdak replied:
There are defaults, that serve documentation purposes mostly, but there is an explicit fail step: 

```
- fail: msg=Please change default password for jenkins user
   when: jenkins_auth_realm.plain_password == 'jenkins'
```
We generally assume that specific settings will be provided using either --extra-vars @file or by other means applicatnle to following setup, like `host_vars` (Ansible is super elastic about that). 

Vagrant playbook attached to this role requires you to pass the var file in an environment  variable.

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+as this sets up the **shell** password for this user. 
+
+You'll need to set both a plain password and a hashed one.
+To obtain hashed password use `mkpasswd` command, for example: 
```
nit: To obtain a hashed password, use the ...

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+
+You'll need to set both a plain password and a hashed one.
+To obtain hashed password use `mkpasswd` command, for example: 
+`mkpasswd --method=sha-512`. (Note: hashed password is required 
```
nit:  Note: a hashed..

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+
+#### Jenkins credentials
+
+Jenkins contains its own credential store, to fill it with credentials 
```
nit:  "... credential store.  To fill it with credentials, please use the ..."

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+Jenkins contains its own credential store, to fill it with credentials 
+please use `JENKINS_ANALYTICS_CREDENTIALS` variable. This variable 
+is a list of objects, each object represents a single credential,
+for now passwords and ssh-keys are supported. 
```
s/represents/representing/.
s/single credential,/single credential./
s/for now/For now,/

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+which should contain contents of private key used for 
+authentication to checkout github repositories.  
+
+Each credential has an unique id, this credential id is used by 
```
s/id, this/id. This/

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+
+Each credential has an unique id, this credential id is used by 
+Jenkins jobs to match a task requiring credential with a 
+credential for given task. 
```
I don't understand this sentence.   Is this accurate:   "to match the credential to the task or tasks for which it is needed." ?

jbzdak replied:
How about this wording: "Each credential has an unique id. A given credential is identified by its id, and that id can be used to by a task to reference a the credential."

bradenmacdonald replied:
"Each credential has **a** unique ID, which is used to match the credential to the task(s) for which it is needed."

(Saying "x is identified by its id" is redundant/tautological. Also, note that the indefinite article before "unique" is "a".)

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+    JENKINS_ANALYTICS_CREDENTIALS:
+      # id is a scope-unique credential identifier
+      - id: test-password
+        # Scope must be global, to have other scopes you'll need ot modify addCredentials.groovy
```
s/global, to/global. To/.     Also s/need ot/need to/.

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+Jenkins auth realm, that is a method used by jenkins to authenticate users.
+Realm type stored in `jenkins_auth_realm.name` variable.
+
+In future we will try enable other auth domains, while
```
try to enable

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+
+#### Auth realm
+
+Jenkins auth realm, that is a method used by jenkins to authenticate users.
```
Jenkins auth realm is a method used by Jenkins ...  ?

jbzdak replied:
I reworded the paragraph to sound better: 

Jenkins auth realm encapsulates user management in Jenkins, that is:

* What users are can log in
* What credentials they use to log in

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+For now only `unix` realm supported -- which requires every user to have a
+shell account on the server.
+
+Unix realm requires following settings:
```
requires the following...

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+
+#### Seed job configuration
+
+Seed job is configured in `jenkins_seed_job` variable, which has following
```
has the following

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+attributes:
+
+* `name`:  Name of the job in Jenkins.
+* `time_trigger`: A Jenkins cron entry defining how often this job is ran.
```
s/is ran/should run/.

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+
+* `name`:  Name of the job in Jenkins.
+* `time_trigger`: A Jenkins cron entry defining how often this job is ran.
+* `removed_job_action`: what to do when job created by the seed job is deleted.
```
when a job

jbzdak replied:
FYI: Reworded this part.

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+* `time_trigger`: A Jenkins cron entry defining how often this job is ran.
+* `removed_job_action`: what to do when job created by the seed job is deleted.
+  This can be either  `DELETE` or`IGNORE`.
+* `removed_view_action`: What to do when view created by the seed job is removed.
```
when a view

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+* `removed_view_action`: What to do when view created by the seed job is removed.
+  This can be either  `DELETE` or`IGNORE`.
+* `scm`: Scm object is used to define seed job repository and related settings.
+  It has following properties:
```
has the following

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+  * `scm.credential_id`: Id of credential used to authenticate to the repository.
+    This setting is optional. If it is missing or falsy, credentials will be omitted. 
+    Please note that when you use ssh repository url, you'll need to set up key irregardless 
+    of whether repository is public or private (to establish an ssh connection
```
whether the repository

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+  * `scm.url`: URL for the repository.
+  * `scm.credential_id`: Id of credential used to authenticate to the repository.
+    This setting is optional. If it is missing or falsy, credentials will be omitted. 
+    Please note that when you use ssh repository url, you'll need to set up key irregardless 
```
s/set up key irregardless of whether/set up a key regardless whether/.

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+  It has following properties:
+  * `scm.type`: It must have value of `git`.
+  * `scm.url`: URL for the repository.
+  * `scm.credential_id`: Id of credential used to authenticate to the repository.
```
credential to use when authenticating

jbzdak replied:
It's not a credential itself, but an id of the credential, corresponding to ``JENKINS_ANALYTICS_CREDENTIALS.*.id``, it's important to underline that this is a pointer.

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+        type: git
+        url: "git@github.com:edx-ops/edx-jenkins-job-dsl.git"
+        credential_id: "github-deploy-key"
+        targed_jobs: "jobs/analytics-edx-jenkins.edx.org/*Jobs.groovy"
```
s/targed_jobs/target_jobs/.

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+      removed_view_action: "IGNORE"
+      scm:
+        type: git
+        url: "git@github.com:edx-ops/edx-jenkins-job-dsl.git"
```
Should we mention edx-ops?

jbzdak replied:
Readme would be very vague without this. AFAIK @e0d  wants to open-source ``git@github.com:edx-ops/edx-jenkins-job-dsl.git`` at some point in time, unless this just needs to be removed, then I'll remove it.

brianhw replied:
I thought it would be cleaner to not make references to non-open-sourced repos from open-source repos.  I had assumed that edx-ops would never had open-sourced repos, and that if we open-sourced the job-dsl repo, that it would be in the edx account.  And that for this level, it would be more general to have "git@github.com:<account>/<jenkins-job-dsl-repo>.git".  But if we are planning to open-source this soon, then I'd be willing to leave in place.

jbzdak replied:
@brianhw If this repository was opaque I'd lean-on adding generic git url, but this repository is not opaque, and this job relies on having there specific dsl jobs, in #2830 we increase this connection between this role and dsl repository. 

If the DSL experiemnt works out really well we might consider refactoring this work to extract a ``jenkins_dsl`` role, where this could point to any repository containing dsl jobs.

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+    Please note that when you use ssh repository url, you'll need to set up key irregardless 
+    of whether repository is public or private (to establish an ssh connection
+    you need a valid public key). 
+  * `scm.target_jobs`: A shell glob expression relative to repo selecting
```
Is this also relative to repo root (like the classpath below)?

jbzdak replied:
Yes, it is. Fixed in comments.

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,202 @@
+   `get-job`, `build <<job>>` commands wouldn't work.
+   Giving anonymous these permissions is a workaround for
+   transient Jenkins issue (reported [couple][1] [of][2] [times][3]).
+3. We force unix authentication method -- that is every user that can login
```
add comma after "that is"

brianhw commented:
Just cosmetic stuff about comments.  A few I wasn't sure I understood; a few other suggestions made as well.  Otherwise :+1:

jbzdak commented:
@brianhw Thank you for the conditional approval, I'll fix all comments shortly.

jbzdak commented:
@bradenmacdonald @e0d @brianhw I have addressed your comments, thank for your insight. 

I have removed the underscored from the begins of file names (`_add_credentials.yaml` -> `add_credentials.yaml`), and I have moved auth  realm files back to the `main.yaml`, however I decided to leave `add_credentials.yaml` and `create_seed_job.yaml` as a separate files, as to me they provide abstraction (they contain non-trivial self-contained logic). If this makes sense we are ready for merge, if not please ping me again and I'll just merge them to `main.yaml`.

brianhw commented on playbooks/roles/jenkins_analytics/tasks/main.yml:
```diff
@@ -0,0 +1,39 @@
+    owner: "{{ jenkins_user }}"
+    group: "{{ jenkins_group }}"
+
+# Unconditionally restart jenkins, this has two side-effects:
```
nit:   s/jenkins/Jenkins/

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,206 @@
+
+Jenkins auth realm encapsulates user management in Jenkins, that is:
+
+* What users are can log in
```
s/users are can/users can/

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,206 @@
+Realm type stored in `jenkins_auth_realm.name` variable.
+
+In future we will try to enable other auth domains, while
+preserving ability to run cli.
```
nit: preserving the ability

brianhw commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,206 @@
+attributes:
+
+* `name`:  Name of the job in Jenkins.
+* `time_trigger`: A Jenkins cron entry defining how often this job should ran.
```
s/is ran/should run/.

brianhw commented:
Thanks for putting up with my nits.   I added a few more, but none are blockers. And I just have one remaining question about how default passwords work.

brianhw commented:
LGTM. :+1:

jbzdak commented:
Every outstanding issue have been taken into account, tests pass, jenkins provisions locally and seed job works. Will merge shortly on @e0d [conditional approval](https://github.com/edx/configuration/pull/2810#issuecomment-197093447). 

CC: @brianhw @bradenmacdonald

e0d commented on playbooks/roles/jenkins_analytics/README.md:
```diff
@@ -0,0 +1,206 @@
+as this sets up the **shell** password for this user. 
+
+You'll need to set both a plain password and a hashed one.
+To obtain a hashed password use the `mkpasswd` command, for example: 
```
Which mkpasswd tool is this?  There doesn't seem to be any command on ubuntu.

arbabnazar replied:
The `whois` package makes the mkpasswd command available on Ubuntu
``` sudo apt-get install -y whois```

jbzdak replied:
Strangely enough it's in the ``whois`` package.
